spectrum/xpub.py

```diff
@@ -0,0 +1,65 @@
+from spectrum import logger
+
+
+LOGGER = logger.logger(__name__)
+
+def _log(message, *args, **kwargs):
+    LOGGER.info(message, extra={'app':'elife-xpub'}, *args, **kwargs)
+
+class PageObject:
+    def __init__(self, driver):
+        self._driver = driver
+
+    def _input(self, css_selector, text, name=None):
+        input_element = self._driver.find_element_by_css_selector(css_selector)
+        if name:
+            _log("Found %s input %s", name, css_selector)
+        input_element.send_keys(text)
+
+
+class XpubJavaScriptSession:
+    CSS_LOGIN_BUTTON = 'button[data-test-id="login"]'
+    CSS_PROFILE_MENU = 'button[data-test-id="profile-menu"]'
+
+    def __init__(self, driver):
+        self._driver = driver
+
+    def login(self):
+        login_button = self._driver.find_element_by_css_selector(self.CSS_LOGIN_BUTTON)
+        _log("Found login button %s `%s`", self.CSS_LOGIN_BUTTON, login_button.text)
+        login_button.click()
+        _log("Clicked login button %s", self.CSS_LOGIN_BUTTON)
+        profile_menu = self._driver.find_element_by_css_selector(self.CSS_PROFILE_MENU)
+        _log("Found profile menu %s", self.CSS_PROFILE_MENU)
+        profile_menu.click()
+        _log("Clicked profile menu %s", self.CSS_PROFILE_MENU)
+
+    def dashboard(self):
+        dashboard_button = self._driver.find_element_by_link_text('Dashboard')
+        dashboard_button.click()
+        return XpubDashboardPage(self._driver)
+
+
+class XpubDashboardPage(PageObject):
+    CSS_NEW_SUBMISSION_BUTTON = 'button[data-test-id="desktop-new-submission"]'
+
+    def create_initial_submission(self):
+        new_submission_button = self._driver.find_element_by_css_selector(self.CSS_NEW_SUBMISSION_BUTTON)
+        new_submission_button.click()
+        return XpubInitialSubmissionAuthorPage(self._driver)
+
+
+class XpubInitialSubmissionAuthorPage(PageObject):
+    CSS_INPUT_FIRST_NAME = 'input[name="author.firstName"]'
+    CSS_INPUT_LAST_NAME = 'input[name="author.lastName"]'
+    CSS_INPUT_EMAIL = 'input[name="author.email"]'
+    CSS_INPUT_AFFILIATION = 'input[name="author.aff"]'
+    CSS_NEXT = 'button[data-test-id="next"]'
+
+    def next(self):
+        self._input(self.CSS_INPUT_FIRST_NAME, 'Josiah', 'first name')
```

Comment:
Using the fluent API approach this would become something like `self.css(self.CSS_INPUT_FIRST_NAME), name='first name').set_input('Josiah')`.

Although really each input should be wrapped as a function or constant if the API supports it (i.e. deferred lookup). Could use a `@property` as well. so that this could become:

```python
@property
def first_name(self):
    return self.css(self.CSS_INPUT_FIRST_NAME), name='first name')

def next(self):
    self.first_name.set_input('Josiah')
    # ...
```

Reply:
Would want to only query for the element once and hold it in memory. Not sure if it's the case to expose the elements publicly, this should be usable without leaking knowledge about Selenium and `python-selenium` types.

Reply:
I guess the text book _Page Object_ has methods like `set_first_name` etc. but you will end up with duplication code that's not necessarily more helpful because you might then add `clear_first_name`. So the idea is to have a _Page Object_ or rather _Component Object_ for an input field. Whether we are creating our own _Input Component Object_ or use a library that already has it and is suitable may be another question. It can be a very simple object with the interactions you might have with the component (_Input_, _Button_, etc.) - exactly like a _Page Object_ just for a _Component_.

To really separate the test from the code you'd use something like _Gherkin_.

But it sounds like we don't actually want to write functional tests but only smoke tests. In that case it's probably overkill.